### PR TITLE
*: support check an index

### PR DESCRIPTION
**Pull request by zimulala**
Using `IndexLoopUpReader` to support `admin check index` option.

In next PR I will handle the case below：
```
index     data (handle, data): (1, 10), (2, 20), (4, 30)
table     data (handle, data): (1, 10), (2, 20), (4, 40)
```

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @coocood @shenli @winoros

File: parser/parser_test.go [hunk 1]
```diff
@@ -403,6 +403,7 @@ func (s *testParserSuite) TestDMLStmt(c *C) {
 		{"admin show ddl;", true},
 		{"admin show ddl jobs;", true},
 		{"admin check table t1, t2;", true},
+		{"admin check index idx;", true},
```
Comment (lamxTyler) [inline parser/parser_test.go]: So the `idx` is the table name, and index name is empty?

File: plan/planbuilder.go [hunk 2]
```diff
@@ -407,13 +408,84 @@ func (b *planBuilder) buildPrepare(x *ast.PrepareStmt) Plan {
...
+
+	// get index information
+	var idx *model.IndexInfo
+	for _, index := range tblInfo.Indices {
+		if index.Name.L == as.Index {
+			idx = index
+			break
+		}
+	}
+	if idx == nil {
```
Comment (lamxTyler) [inline plan/planbuilder.go]: Do we need to check if the index state is public?

File: executor/executor.go [hunk 3]
```diff
@@ -409,6 +409,70 @@ func (e *CheckTableExec) run(ctx context.Context) error {
...
+}
+
+// NextChunk implements the Executor NextChunk interface.
+func (e *CheckIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) error {
+	if e.done {
+		return nil
+	}
+	err := e.run(ctx)
+	e.done = true
+	return errors.Trace(err)
```
Comment (lamxTyler) [inline executor/executor.go]: Can we just call `e.Next` or extract a function?

File: executor/executor.go [hunk 4]
```diff
@@ -409,6 +409,70 @@ func (e *CheckTableExec) run(ctx context.Context) error {
...
+}
+
+func (e *CheckIndexExec) run(ctx context.Context) error {
+	err := admin.CheckIndicesCount(e.ctx, e.dbName, e.tableName, []string{e.idxName})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for {
+		row, err := e.src.Next(ctx)
```
Comment (winoros) [inline executor/executor.go]: Use `NextChunk`?

File: executor/executor.go [hunk 5]
```diff
@@ -409,6 +409,70 @@ func (e *CheckTableExec) run(ctx context.Context) error {
...
+// CheckIndexExec represents the executor of check an index.
+// It is built from the "admin check index" statement, and it checks if the
+// index matches the records in the table.
+type CheckIndexExec struct {
+	baseExecutor
+
+	dbName    string
+	tableName string
+	idxName   string
+	src       Executor
```
Comment (zz-jason) [inline executor/executor.go]: s/Executor/*IndexLookUpExecutor/ is more readable ?

File: executor/distsql.go [hunk 6]
```diff
@@ -878,6 +886,10 @@ func (w *tableWorker) executeTask(ctx context.Context, task *lookupTableTask) {
 		}
 		sort.Sort(task)
 	}
+
+	if w.isCheckOp && handleCnt != len(task.rows) {
+		err = errors.Errorf("handle count %d isn't equal to value count %d", handleCnt, len(task.rows))
```
Comment (shenli) [inline executor/distsql.go]: Add the missing handle and corresponding index value in the error message.

File: executor/executor.go [hunk 7]
```diff
@@ -409,6 +409,70 @@ func (e *CheckTableExec) run(ctx context.Context) error {
 	return nil
 }
 
+// CheckIndexExec represents the executor of check an index.
```
Comment (shenli) [inline executor/executor.go]: check -> checking

File: executor/executor.go [hunk 8]
```diff
@@ -409,6 +409,70 @@ func (e *CheckTableExec) run(ctx context.Context) error {
...
+// It is built from the "admin check index" statement, and it checks if the
```
Comment (shenli) [inline executor/executor.go]: it checks the consistency of the index data with the records of the table.

File: plan/planbuilder.go (hunk 2 above)
Comment (zimulala) [reply plan/planbuilder.go]: Done.

**Comment by zimulala** [comment]
PTAL @lamxTyler @winoros @zz-jason @shenli

File: executor/executor.go [hunk 9]
```diff
@@ -409,6 +409,81 @@ func (e *CheckTableExec) run(ctx context.Context) error {
...
+	if e.done {
+		return nil
+	}
+	defer func() { e.done = true }()
+
+	err := admin.CheckIndicesCount(e.ctx, e.dbName, e.tableName, []string{e.idxName})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cnt := 0
```
Comment (zz-jason) [inline executor/executor.go]: `cnt` is unused ?

**Comment by zimulala** [comment]
PTAL @zz-jason

**Comment by shenli** [comment]
LGTM

**Pull request merged**